illumos-utils/src/zfs.rs

```diff
@@ -493,6 +511,204 @@ pub struct DatasetEnsureArgs<'a> {
     pub additional_options: Option<Vec<String>>,
 }
 
+// If this dataset will have a well-defined mountpoint, we ensure an immutable directory exists
+// here.
+//
+// This is intended to mitigate issues like:
+// - https://github.com/oxidecomputer/omicron/issues/7874
+// - https://github.com/oxidecomputer/omicron/issues/4203
+//
+// Suppose we have a dataset, named "crypt/debug", within a parent dataset named "crypt".
+//
+// Suppose "crypt" is mounted at "/storage/crypt", and we want to mount "crypt/debug" as
+// "/storage/crypt/debug".
+//
+// When we create the debug dataset, we are creating a directory within the "crypt" dataset, and
+// mounting our new dataset on top of it. Without modification, this presents a threat: the
+// "/storage/crypt/debug" directory is a regular directory, and if the underlying dataset is
+// unmounted (explicitly, or due to a reboot) data could be placed UNDERNEATH the mount point.
+//
+// To mitigate this issue, we create the mountpoint ahead-of-time, and set the immutable
+// property on it. This prevents the mountpoint from being used as anything other than a
+// mountpoint.
+//
+// NOTE: This must only be called on mountpoints for datasets which ARE NOT MOUNTED.
```

Comment:
What happens if this is called and the dataset _is_ mounted? Do we need to guard against that? 

Edit: after reading below, I _think_ if we did this, we'd make the mounted directory immutable, make a temporary directory, then fail in the `rename()` calls (since rename won't cross filesystems, right?). The renames failing is good, but making the mounted directory immutable seems pretty bad...